### DISCUSSION TITLE
GLMakie lines TODO

### DISCUSSION THREAD

**Discussion by ffreyer** [JuliaPlots/Makie.jl#4230] [General]
As I'm working on #2536 I keeping finding little issues and things we should add or change

1. ~~Solid lines are only anti-aliased along their width, but not their length (i.e. the ends aren't anti-aliased)~~ Fixed by #2666
2. Solid (probably all) lines overlap themselves when corners are sharp enough
3. Since I'm adding linecaps in #2536 we should also add joint styles. Options should include round, sharp corner via projecting (we do this for small changes in line direction), cut off corners (we do this for large changes in direction) and the "smart" version we have atm
4. `poly` picks the first method definition https://github.com/MakieOrg/Makie.jl/blob/1c6d934dd279466db4f06c897953909b8b10ceee/src/basic_recipes/poly.jl#L11 for `Rect`, which means it relies on `wireframe` which uses `linesegments`. So it will always produce bad corners (fully overlapping, partially overlapping or missing corners). We should make it go into the second method https://github.com/MakieOrg/Makie.jl/blob/1c6d934dd279466db4f06c897953909b8b10ceee/src/basic_recipes/poly.jl#L86 to produce at least 3 clean corners via `lines`
5. We should add a way to draw closed lines to avoid the last overlapping corner. This might require some adjustments to joints, so do it with (3)
6. ~~Non-solid lines sometimes produce artifacts~~ Fixed by #2666
7. ~~Non-solid lines have irregular spacing, especially on curved lines~~ Fixed by #2666
8. Lines produce artifacts on other objects with `fxaa = true`
9. ~~A very short line creates downward artifacts on a 90° turn~~ Duplicate of #1129, improved by #2666, and further improved by #2843

## Example (1), (2)

This is the miter joints test image from GLMakie

![Screenshot from 2023-01-03 17-25-11](https://user-images.githubusercontent.com/10947937/210398375-f5bca919-6b65-4d5a-8280-fd2e359d953f.png)

## Example (4)

On master:
```julia
scene = Scene()
poly!(scene, Rect2f(-0.8, -0.8, 1.6, 1.6), color = :transparent, strokewidth = 10, strokecolor = (:black, 0.5))
```

![Screenshot from 2023-01-03 17-47-33](https://user-images.githubusercontent.com/10947937/210402412-ce22f331-4e07-484b-929e-de92eec503e8.png)

## Example (6)

On master. These artifacts get more or less pronounced when the line is moved around.
```julia
scene = Scene()
lines!(scene, [-1,0,1], [0,0,0], linestyle = :dot, linewidth = 20)
```
![Screenshot from 2023-01-03 17-42-06](https://user-images.githubusercontent.com/10947937/210401726-bb0cbc1e-4db6-4486-af10-f03324b1525e.png)

## Example (7)

This is from the linecaps pr, which is already better than master but still has issues with this
```julia
lines(range(0, 2pi, length=31), sin, linestyle = :dash, linewidth=5)
```

![Screenshot from 2023-01-02 13-45-05](https://user-images.githubusercontent.com/10947937/210400866-b6b4e010-4f05-4950-ba55-3a97909853f4.png)

## Example (8)

This is from a random `crossbar` with the center line drawn

![Screenshot from 2023-01-03 17-36-16](https://user-images.githubusercontent.com/10947937/210400466-b6f6cce8-6bc1-4b49-b9c9-8ed829f1b823.png)

## Example (9)

On master this generates artifacts depending on the y zoom level

```julia
lines([0, 1, 2, 2, 0], [0, 1, 0.01, 0, 0], linewidth = 10)
```

![Screenshot from 2023-01-03 19-55-17](https://user-images.githubusercontent.com/10947937/210424032-2ee20655-fb39-475d-9ee0-701ecffff399.png) ![Screenshot from 2023-01-03 19-55-23](https://user-images.githubusercontent.com/10947937/210424036-66bcbd82-d249-4330-83ef-a8e06a10b6de.png)

**Comment by jkrumbiegel** [comment]
> We should add a way to draw closed lines to avoid the last overlapping corner

+1 for that from me, currently in CairoMakie this is just done if the last point of a line is equal to the first. I just decided to go with that at the time even though it's not "official" behavior because too often it's what should happen anyway.

**Comment by SimonDanisch** [comment]
How much is fixed of this now?

**Comment by ffreyer** [comment]
What I crossed out in the list should still be up to date

**Comment by ffreyer** [comment]
I think the self-overlap (2) isn't something we can fix without another full rewrite. (Possibly switching to a two pass line rendering solution, where we render line opaque first, so the overlap doesn't matter and apply transparency in a second step when merging with the main buffer.)

I had some notes on closed lines (5) here: https://github.com/MakieOrg/Makie.jl/pull/2666#issuecomment-1434807798 Should be relatively easy, but requires adding extra points.

(8) requires a different fxaa solution which I don't really see happening anytime soon.

**Comment by ffreyer** [comment]
Closing in favor of #4262
